src/Workspaces/Remote/ServiceHub/Host/RemoteWorkspace.SolutionCreator.cs

```diff
@@ -115,105 +113,147 @@ public async Task<Solution> CreateSolutionAsync(Checksum newSolutionChecksum, Ca
             private async Task<Solution> UpdateProjectsAsync(
                 Solution solution, SolutionStateChecksums oldSolutionChecksums, SolutionStateChecksums newSolutionChecksums, CancellationToken cancellationToken)
             {
-                using var olds = SharedPools.Default<HashSet<Checksum>>().GetPooledObject();
-                using var news = SharedPools.Default<HashSet<Checksum>>().GetPooledObject();
+                var solutionState = solution.SolutionState;
 
-                olds.Object.UnionWith(oldSolutionChecksums.Projects.Checksums);
-                news.Object.UnionWith(newSolutionChecksums.Projects.Checksums);
+                using var _1 = PooledDictionary<ProjectId, Checksum>.GetInstance(out var oldProjectIdToChecksum);
+                using var _2 = PooledDictionary<ProjectId, Checksum>.GetInstance(out var newProjectIdToChecksum);
+                using var _3 = PooledHashSet<ProjectId>.GetInstance(out var allProjectIds);
 
-                // remove projects that exist in both side
-                olds.Object.ExceptWith(newSolutionChecksums.Projects.Checksums);
-                news.Object.ExceptWith(oldSolutionChecksums.Projects.Checksums);
+                foreach (var (oldChecksum, projectId) in oldSolutionChecksums.Projects)
+                    oldProjectIdToChecksum.Add(projectId, oldChecksum);
 
-                using var _1 = PooledDictionary<ProjectId, ProjectStateChecksums>.GetInstance(out var oldMap);
-                using var _2 = PooledDictionary<ProjectId, ProjectStateChecksums>.GetInstance(out var newMap);
+                foreach (var (newChecksum, projectId) in newSolutionChecksums.Projects)
+                    newProjectIdToChecksum.Add(projectId, newChecksum);
```

Comment:
done as explicit for loops to prevent allocations.

Reply:
note: the corresponding 'Except' work moved lower.

Reply:
> done as explicit for loops to prevent allocations.

If you're doing that for perf add a comment before somebody tries to "optimize" this later.